ansible_runner/streaming.py

```diff
@@ -93,62 +159,70 @@ def update_paths(self, kwargs):
-
-        self.kwargs.update(self.job_kwargs)
-        self.kwargs['quiet'] = True
-        self.kwargs['suppress_ansible_output'] = True
-        self.kwargs['private_data_dir'] = self.private_data_dir
-        self.kwargs['status_handler'] = self.status_handler
-        self.kwargs['event_handler'] = self.event_handler
-        self.kwargs['artifacts_handler'] = self.artifacts_handler
-        self.kwargs['finished_callback'] = self.finished_callback
-
-        r = ansible_runner.interface.run(**self.kwargs)
-        self.status, self.rc = r.status, r.rc
 
-        # FIXME: do cleanup on the tempdir
+                if 'kwargs' in data:
+                    self.job_kwargs = self.update_paths(data['kwargs'])
+                elif 'zipfile' in data:
+                    try:
+                        unstream_dir(self._input, data['zipfile'], self.private_data_dir)
+                    except Exception:
+                        self.status_handler({
+                            'status': 'error',
+                            'job_explanation': 'Failed to extract private data directory on worker.',
+                            'result_traceback': traceback.format_exc()
+                        }, None)
+                        self.finished_callback(None)  # send eof line
+                        return self.status, self.rc
+                elif 'eof' in data:
+                    break
+
+            self.kwargs.update(self.job_kwargs)
+            self.kwargs['quiet'] = True
+            self.kwargs['suppress_ansible_output'] = True
+            self.kwargs['private_data_dir'] = self.private_data_dir
+            self.kwargs['status_handler'] = self.status_handler
+            self.kwargs['event_handler'] = self.event_handler
+            self.kwargs['artifacts_handler'] = self.artifacts_handler
+            self.kwargs['finished_callback'] = self.finished_callback
+
+            r = ansible_runner.interface.run(**self.kwargs)
+            self.status, self.rc = r.status, r.rc
+
+            # FIXME: do cleanup on the tempdir
+        finally:
+            self._end_keepalive()
 
         return self.status, self.rc
 
+    @_synchronize_output_reset_keepalive
     def status_handler(self, status_data, runner_config):
         self.status = status_data['status']
         self._output.write(json.dumps(status_data).encode('utf-8'))
         self._output.write(b'\n')
         self._output.flush()
 
+    @_synchronize_output_reset_keepalive
     def event_handler(self, event_data):
         self._output.write(json.dumps(event_data).encode('utf-8'))
         self._output.write(b'\n')
         self._output.flush()
 
+    @_synchronize_output_reset_keepalive
     def artifacts_handler(self, artifact_dir):
         stream_dir(artifact_dir, self._output)
         self._output.flush()
 
+    @_synchronize_output_reset_keepalive
     def finished_callback(self, runner_obj):
         self._output.write(json.dumps({'eof': True}).encode('utf-8'))
```

Comment:
you'd probably want to do `self._end_keepalive()` before this line, because eof should be the final line. Wouldn't break anything I know of to have this violated, but better not risk it. If so, this method shouldn't need the decorator either.

Reply:
Since `_end_keepalive()` doesn't currently `join`, it's probably best to keep this one decorated for the lock (since the keepalive thread could still splat something on its dying gasp). That said, I think it *would* be smart to add a defensive call to `_end_keepalive()` inside this to guarantee that it's definitively impossible to splat a keepalive after the eof message- I'll add that.

Reply:
That answered my question completely. But raised a new question with

> `_end_keepalive()` doesn't currently `join`

Couldn't something go wrong with that? If something was using `Worker` programmatically, will this leave around useless threads? Particularly with multiple calls. While this doesn't seem like a very realistic concern, it's already a concern for ansible-runner tests.

My current reading is that if the parameter `keepalive_seconds` parameter gets its default value of `0`, this will still start a thread that will never time out (but still gets reset, kind of pointlessly, by other reads).

This is still a draft, and I'm not asking you to do the work, but want general oppinions on:
 - If the parameter is 0, could we avoid creating an extra thread entirely, minimally affecting the current code path (which all AWX jobs hit)?
 - For test safety (if nothing else) could we have a means of joining the thread? I might agree that never joining is best for CLI use, but a programmatic hook for testing might be useful.

Reply:
> could we avoid creating an extra thread entirely

Yeah, I actually wrote it that way to begin with (as well as no-opping the event and `nullcontext`ing the lock), but since the overhead is relatively low, I chucked that since fewer code path permutations and optimizations is much easier to test (especially where concurrency is involved).

> if the parameter keepalive_seconds parameter gets its default value of 0, this will still start a thread that will never time out 

[this](https://github.com/ansible/ansible-runner/pull/1191/files#diff-81c3df3f520280c8548b6499f287446e289a81ecad30288e0c561d3aa95456c5R105) causes the thread to exit instantly in that case.

I was back-and-forth on the `join()` as well- since I pulse the event in the shutdown request anyway, in *nearly* all cases, the thread should exit instantly, but there is a non-zero possibility that it could wait one keepalive interval. Which does bring up the point that I should mark that thread `daemon=True`. (it was originally a futures TPE, but that seemed like more overhead than necessary for the fire-and-forget, esp since we can't just bridge the `cancel` on the future to actually function :( )

Reply:
* added the defensive `_end_keepalive` call to the finished callback, just in case
* the keepalive thread state is already inspectable by tests via the worker's `_keepalive_thread` attribute
* we can cheaply short-circuit the creation of the thread entirely in `_begin_keepalive()` if we want; currently it just immediately exits the thread if KA is disabled